README.md
Vantiv eCommerce PHP SDK
=====================
#### WARNING:
##### All major version changes require recertification to the new version. Once certified for the use of a new version, Vantiv modifies your Merchant Profile, allowing you to submit transaction to the Production Environment using the new version. Updating your code without recertification and modification of your Merchant Profile will result in transaction declines. Please consult you Implementation Analyst for additional information about this process.
About Vantiv eCommerce
------------
[Vantiv eCommerce](https://developer.vantiv.com/community/ecommerce) powers the payment processing engines for leading companies that sell directly to consumers through  internet retail, direct response marketing (TV, radio and telephone), and online services. Vantiv eCommerce is the leading authority in card-not-present (CNP) commerce, transaction processing and merchant services.


About this SDK
--------------
The Vantiv eCommerce PHP SDK is a PHP implementation of the [Vantiv eCommerce](https://developer.vantiv.com/community/ecommerce). XML API. This SDK was created to make it as easy as possible to connect process your payments with Vantiv eCommerce.  This SDK utilizes  the HTTPS protocol to securely connect to Vantiv eCommerce.  Using the SDK requires coordination with the Vantiv eCommerce team in order to be provided with credentials for accessing our systems.

Each PHP SDK release supports all of the functionality present in the associated Vantiv eCommerce XML version (e.g., SDK v9.3.2 supports Vantiv eCommerce XML v9.3). Please see the online copy of our XSD for Vantiv eCommerce XML to get more details on what the Vantiv eCommerce payments engine supports.

This SDK was implemented to support the PHP programming language and was created by Vantiv eCommerce. Its intended use is for online transaction processing utilizing your account on the Vantiv eCommerce payments engine.

See LICENSE file for details on using this software.

Source Code available from : https://github.com/Vantiv/cnp-sdk-for-php

Please contact [Vantiv eCommerce](https://developer.vantiv.com/community/ecommerce) to receive valid merchant credentials in order to run tests successfully or if you require assistance in any way.  We are reachable at sdksupport@Vantiv.com

SDK PHP Dependencies
--------------
Up to date list available at [Packagist](https://packagist.org/packages/litle/payments-sdk)

Setup
============
Using with composer
--------------------
If you are using a composer to manage your dependencies, you can do the following in your project directory:

1) Install the composer using command:
> curl -sS https://getcomposer.org/installer | php

2) Install dependencies using the command:
> php composer.phar install

3) Configure the SDK:
> cd cnp/sdk
> php Setup.php

4) Run the attached sample:
```php
<?php
require_once _DIR_.'/vendor/autoload.php';
#sale
$sale_info = array(
             'orderId' => '1',
             'amount'  => '10010',
             'orderSource' => 'ecommerce',
             'billToAddress' => array(
             'name' => 'John Smith' ,
             'addressLine1' => ' 1 Main St.',
             'city' => 'Burlington' ,
             'state' => 'MA' ,
             'zip' => '0183-3747',
             'country' => 'US'),
             'card' => array(
             'number' => '5112010000000003',
             'expDate' => '0112',
             'cardValidationNum' => '349',
             'type' => 'MC' )
            );
$initialize = new cnp\sdk\CnpOnlineRequest();
$saleResponse =$initialize->saleRequest($sale_info);
#display results
echo ("Response: " . (cnp\sdk\XmlParser::getNode($saleResponse,'response')) . "<br>");
echo ("Message: " . cnp\sdk\XmlParser::getNode($saleResponse,'message') . "<br>");
echo ("Vantiv eCommerce Transaction ID: " . cnp\sdk\XmlParser::getNode($saleResponse,'cnpTxnId'));
```
> php your_sample_name

Using without composer
-----------------------
If you're not, you have to add a require for each and every class that's going to be used.

1) Configure the SDK
> cd into cnp/sdk
> php Setup.php

2) Add the cnp folder and require the path for your file

3) run your file 

> php your_file

Clone Repo
---------------

1) Install the Vantiv eCommerce PHP SDK from git. 

> git clone git://github.com/Vantiv/cnp-sdk-for-php.git

> php ~/composer.phar install


2) Once the SDK is downloaded run our setup program to generate a configuration file.

> cd cnp-sdk-for-php/lib

> php Setup.php

Running the above commands will create a configuration file in the lib directory. 


3) Create a symlink to the SDK

>ln -s /path/to/sdk /var/www/html/nameOfLink


4.) Create a php file similar to: 

```php
<?php
require_once realpath(dirname(__FILE__)) . '/../lib/CnpOnline.php';  
    // Visa $10.00 sale
    $hash_in = array(
	      'amount'=>'106',
	      'orderId' => '123213',
	      'orderSource'=>'ecommerce',
	      'card'=>array(
	      'type'=>'VI',
	      'number' =>'4100000000000001',
	     'expDate' =>'1000')
	      );
//Perform the transaction on the Vantiv eCommerce Platform
$initialize = new CnpOnlineRequest();
$saleResponse = $initialize->saleRequest($hash_in);

// Display Result 
echo ("Message: " . XMLParser::getNode($saleResponse,'message') . "<br>");
echo ("Vantiv eCommerce Transaction ID: " . XMLParser::getNode($saleResponse,'cnpTxnId'));
```

As of 8.13.1, you may also use a tree-oriented style to get the response values:
```php
<?php
require_once realpath(dirname(__FILE__)) . '/../lib/CnpOnline.php';  
    // Visa $10.00 sale
    $hash_in = array(
	      'amount'=>'106',
	      'orderId' => '123213',
	      'orderSource'=>'ecommerce',
	      'card'=>array(
	      'type'=>'VI',
	      'number' =>'4100000000000001',
	     'expDate' =>'1000')
	      );
//Perform the transaction on the Vantiv eCommerce Platform
$initialize = new CnpOnlineRequest($treeResponse=true);
$saleResponse = $initialize->saleRequest($hash_in);

// Display Result 
echo ("Message: " . $saleResponse->saleResponse->message . "<br>");
echo ("Vantiv eCommerce Transaction ID: " . $saleResponse->saleResponse->cnpTxnId);
```


NOTE: you may have to change the path to match that of your filesystems.  

If you get an error like:
```bash
PHP Fatal error:  require_once(): Failed opening required '/home/gdake/git/cnp-sdk-for-php/../lib/CnpOnline.php' (include_path='.:/usr/share/pear:/usr/share/php') in /home/gdake/git/cnp-sdk-for-php/foo.php on line 2
```
You need to change the second line of your script to load the real location of CnpOnline.php

If you get an error like:
```bash
PHP Fatal error:  require(): Failed opening required '/home/gdake/cnp-sdk-for-php/lib/../vendor/autoload.php' (include_path='.:/usr/share/php:/usr/share/pear') in /home/gdake/cnp-sdk-for-php/lib/CnpOnline.php on line 42
```
You probably had a problem with composer.  You can safely remove line 42 if you are not using batch processing, or you can edit it to point at our dependencies that you've downloaded in another way.

5) Next run this file using php on the command line or inside a browser. You should see the following result provided you have connectivity to the Vantiv eCommerce certification environment.  You will see an HTTP error if you don't have access to the Vantiv URL

    Message: Valid Format
    Vantiv eCommerce Transaction ID: <your-numeric-txn-id>

More examples can be found here [php Gists])(https://gist.github.com/litleSDK)

Please contact Vantiv eCommerce Inc. with any further questions.   You can reach us at SDKSupport@Vantiv.com


CHANGELOG
= CnpOnline CHANGELOG

==Version 12.5.0 (Oct 11, 2018)
* Feature: Full functions for XML v12.5
* Feature: add a new type: encryptedCcAccountNumberType
* Feature: add support for encryption on registerTokenRequestType
* Feature: add paymentAccountReferenceNumber in authorizationResponse and saleResponse
* Feature: add disbursementTypeEnum in fastAccessFunding

==Version 12.3.0 (May 14, 2018)
* Feature: Added support for lodging information in authorization, sale, capture, captureGivenAuth, forceCapture and credit
* Feature: Added translateToLowValueToken transaction type
* Feature: Replaced routingPreference with pinlessDebitRequest in sale
* Feature: Added support for showStatusOnly in queryTransaction
* Feature: Added eventType, accountLogin and accountPasshash to fraudCheck
* Bug Fix: Added createDiscount, updateDiscount, deleteDiscount, createAddOn, updateAddOn and deleteAddOn 
           fields to recurring transactions.

==Version 12.1.0 (March 26, 2018)
* Feature: Added support for Visa card on file transactions
* Feature: Added fastAccessFunding transaction

==Version 12.0.1 (March 16, 2018)
* Feature: PGP support for batches

==Version 11.0.1 (May 4, 2017)
* Feature: update phpseclib to v2.0.4

==Version 11.0 (Apr 10, 2017)
* Feature: implement Vantiv eCommerce XMLv11.0
* Feature: SEPA Support
* Feature: iDEAL support
* Feature: new giftCard transactions
* Feature: giroPay support
* Feature: Network Enhancements mandates
* Feature: add support for raw network response
* Bug fix: specify char encoding in HTTP header

==Version 9.3.3
Update SSL version to 6

==Version 9.3.2
HTTP timeout set to 500ms

==Version 9.3.1 (March 9, 2015)
*Feature: PFIF instruction transaction support was added

== Version 9.03.0 (February 17, 2015)
* Feature: Applepay support was added
* Feature: Secondary amount support was added
* Feature: Add support for new Batch transaction in version 9.3 schema

==version 8.29.0 (January 22,2014)
* Feature: Apple pay and secondary amounts now available.

==version 8.27.0 (Auguest 26,2014)
* Feature: Added samples for each type of transactions

== version 8.25.0 (March 7, 2014)
* Feature: Added support for ROAM: Authorization, ForceCapture, CaptureGivenAuth, Sale, Credit

== version 8.24.0 (February 13, 2014)
* Feature: add triggered rules for advancedFraudCheckResult

== Version 8.23.0 (December 18, 2013)
* Feature: advancedFraudChecks now available in authorization and sale 
* Feature: catLevel (Cardholder Activated Terminal) now available in pos
* Feature: advancedFraudResults now available in fraudResult

== Version 8.22.0 (December 18, 2013)
* Feature: Support for the following new gift card transactions:
ActivateReversal, DeactivateReversal, LoadReversal, RefundReversal,
UnloadReversal, DepositReversal
* Feature: UpdateSubscription now can take token or paypage updates, and
can return token in the response
* Feature: Support for virtualGiftCard in giftCard activations and can be returned in EnhancedAuthResponses
* Feature: Gift Card responses can be returned from AuthReversal

== Version 8.21.0 (December 18, 2013)
* Feature: Support for the following new recurring transactions:
CreatePlan, UpdatePlan
* Feature: Support for the following new gift card transactions:
Activate, Deactivate, Load, Unload, Balance Inquiry
* Feature: Gift card responses are now returned as part of
authorzationResponse, authReversalResponse, captureResponse,
forceCapture, captureGivenAuthResponse, saleResponse, creditResponse
* Feature: fraudResult is now returned as part of captureResponse,
forceCaptureResponse, captureGivenAuthResponse, creditResponse

== Version 8.20.0 (December 17, 2013)
* Feature: More initial support for Recurring, including the ability to update or cancel subscriptions
* Cleanup: Move phpseclib to composer

== Version 8.19.0 (July 30, 2013)
* Feature: Added deptRepayment as a choice for authorizations, forceCapture, captureGivenAuth, sale
* Feature: More initial support for Recurring, including a backwards incompatible changing a field from numberOfPaymentsRemaining to numberOfPayments
* Feature: Add merchantData to forceRequest and captureGivenAuth

== Version 8.18.0 (July 15, 2013)
* Feature: Add initial support for recurring to sales

== Version 8.17.0 (March 29, 2013)
* Feature: Add recyling to voidResponse
* Feature: Add support for surcharging credit card transactions
* Feature: Add support for specifying the terminalId on a pos transaction
* Feature: Add support for pos transactions on tied refunds

== Version 8.15.0 (January 31, 2013)

* Feature: Add merchantData to echeck verifications and echeck redeposits

== Version 8.14.0 (January 30, 2013)

* Feature: Merged pull request #12 from iToto to make convert DOMDocument to string for debugging
* Bugfix: Merged pull request #14 from johnholden to truncate strings that are too long, thus preventing some xml validation errors
* Feature: Temporary storage of card validation num on tokens
* Feature: Updating card validation numbers on tokens

== Version 8.13.1 (June 15, 2012)

* Feature: Add support for accessing response as a tree using simplexml.  Access is provided if $treeResponse=true is passed to the LitleOnlineRequest constructor

== Version 8.13 (May 18, 2012)

* Feature: Authorizations and Sales can accept an optional fraudFilterOverride parameter

== Version 8.12.2 (May 8, 2012)

* Bugfix: Add support for line item data and tax detail to enhanced data

== Version 8.12.1 (April 23, 2012)

* Bugfix: Add support for id, customerid and reportgroup

== Version 8.12 (April 17, 2012)

* Bugfix: Add support for MerchantData on auth and sale
* Feature: Add support for actionReason on credit
* Feature: Track SDK Usage

== Version 8.10.0 (April 2, 2012)

* Initial release


cnp/sdk/Obj2xml.php
<?php
/*
 * Copyright (c) 2011 Vantiv eCommerce Inc.
*
* Permission is hereby granted, free of charge, to any person
* obtaining a copy of this software and associated documentation
* files (the "Software"), to deal in the Software without
* restriction, including without limitation the rights to use,
* copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the
* Software is furnished to do so, subject to the following
* conditions:
*
* The above copyright notice and this permission notice shall be
* included in all copies or substantial portions of the Software.
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND
* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
* WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
* OTHER DEALINGS IN THE SOFTWARE.
*/
namespace cnp\sdk;
require_once realpath(dirname(__FILE__)) . '/CnpOnline.php';
class Obj2xml
{
    public static function toXml($data, $hash_config, $type, $rootNodeName = 'cnpOnlineRequest')
    {
        $config= Obj2xml::getConfig($hash_config, $type);
        $xml = simplexml_load_string("<?xml version='1.0' encoding='utf-8'?><$rootNodeName />");
        $xml-> addAttribute('merchantId',$config["merchantId"]);
        $xml-> addAttribute('version',CURRENT_XML_VERSION);
        $xml-> addAttribute('merchantSdk',$data['merchantSdk']);
        unset($data['merchantSdk']);
        if (isset($data['loggedInUser'])) {
            $xml->addAttribute('loggedInUser',$data["loggedInUser"]);
        };
        unset($data['loggedInUser']);
        $xml-> addAttribute('xmlns:xmlns','http://www.vantivcnp.com/schema');// does not show up on browser docs
        $authentication = $xml->addChild('authentication');
        $authentication->addChild('user',$config["user"]);
        $authentication->addChild('password',$config["password"]);
        $transacType = $xml->addChild($type);
        if (isset($data['partial'])) {
            $transacType-> addAttribute('partial',$data["partial"]);
        };
        unset($data['partial']);
        if (isset($data['customerId'])) {
            $transacType-> addAttribute('customerId',$data["customerId"]);
        };
        unset($data['customerId']);
        if (isset($config['reportGroup'])) {
            $transacType-> addAttribute('reportGroup',$config["reportGroup"]);
        };
        if (isset($data['id'])) {
        if ($data['id'] === "REQUIRED") {
                throw new \InvalidArgumentException("Missing Required Field: id");
            }
            else {
        	$transacType-> addAttribute('id',$data["id"]);
            }
        };
        unset($data['id']);

        Obj2xml::iterateChildren($data,$transacType);

        return $xml->asXML();
    }

    public static function transactionShouldHaveReportGroup($transactionType)
    {
          $transactionsThatDontHaveReportGroup = array(
            'updateSubscription',
            'cancelSubscription',
            'createPlan',
            'updatePlan'
        );

        return (FALSE === array_search($transactionType, $transactionsThatDontHaveReportGroup));
    }

    public static function transactionToXml($data, $type, $report_group)
    {
        $transac = simplexml_load_string("<$type />");
        if (Obj2xml::transactionShouldHaveReportGroup($type)) {
            $transac->addAttribute('reportGroup', $report_group);
            if (isset($data['id'])) {
            	if ($data['id'] === "REQUIRED") {
            		throw new \InvalidArgumentException("Missing Required Field: id");
            	}
            	else {
            		$transac-> addAttribute('id',$data["id"]);
            	}
            };
            unset($data['id']);
        }
       
        Obj2xml::iterateChildren($data,$transac);

        return str_replace("<?xml version=\"1.0\"?>\n", "", $transac->asXML());
    }

    public static function rfrRequestToXml($hash_in)
    {
        $rfr = simplexml_load_string("<RFRRequest />");
        if (isset($hash_in['cnpSessionId'])) {
            $rfr->addChild('cnpSessionId', $hash_in['cnpSessionId']);
        } elseif (isset($hash_in['merchantId']) && isset($hash_in['postDay'])) {
            $auFileRequest = $rfr->addChild('accountUpdateFileRequestData');
            $auFileRequest->addChild('merchantId', $hash_in['merchantId']);
            $auFileRequest->addChild('postDay', $hash_in['postDay']);
        } else {
            throw new \RuntimeException('To add an RFR Request, either a cnpSessionId or a merchantId and a postDay must be set.');
        }

        return str_replace("<?xml version=\"1.0\"?>\n", "", $rfr->asXML());
    }

    public static function generateBatchHeader($counts_and_amounts)
    {
        $config= Obj2xml::getConfig(array());

        $xml = simplexml_load_string("<batchRequest />");
        $xml->addAttribute('merchantId', $config['merchantId']);
        $xml->addAttribute('merchantSdk', CURRENT_SDK_VERSION);

        $xml->addAttribute('authAmount', $counts_and_amounts['auth']['amount']);
        $xml->addAttribute('numAuths', $counts_and_amounts['auth']['count']);

        $xml->addAttribute('saleAmount', $counts_and_amounts['sale']['amount']);
        $xml->addAttribute('numSales', $counts_and_amounts['sale']['count']);

        $xml->addAttribute('creditAmount', $counts_and_amounts['credit']['amount']);
        $xml->addAttribute('numCredits', $counts_and_amounts['credit']['count']);
        
        $xml->addAttribute('giftCardCreditAmount', $counts_and_amounts['giftCardCredit']['amount']);
        $xml->addAttribute('numGiftCardCredits', $counts_and_amounts['giftCardCredit']['count']);

        $xml->addAttribute('numTokenRegistrations', $counts_and_amounts['tokenRegistration']['count']);

        $xml->addAttribute('numTranslateToLowValueTokenRequests', $counts_and_amounts['translateToLowValueTokenRequest']['count']);

        $xml->addAttribute('captureGivenAuthAmount', $counts_and_amounts['captureGivenAuth']['amount']);
        $xml->addAttribute('numCaptureGivenAuths', $counts_and_amounts['captureGivenAuth']['count']);

        $xml->addAttribute('forceCaptureAmount', $counts_and_amounts['forceCapture']['amount']);
        $xml->addAttribute('numForceCaptures', $counts_and_amounts['forceCapture']['count']);

        $xml->addAttribute('authReversalAmount', $counts_and_amounts['authReversal']['amount']);
        $xml->addAttribute('numAuthReversals', $counts_and_amounts['authReversal']['count']);
        
        $xml->addAttribute('giftCardAuthReversalOriginalAmount', $counts_and_amounts['giftCardAuthReversal']['amount']);
        $xml->addAttribute('numGiftCardAuthReversals', $counts_and_amounts['giftCardAuthReversal']['count']);

        $xml->addAttribute('captureAmount', $counts_and_amounts['capture']['amount']);
        $xml->addAttribute('numCaptures', $counts_and_amounts['capture']['count']);
        
        $xml->addAttribute('giftCardCaptureAmount', $counts_and_amounts['giftCardCapture']['amount']);
        $xml->addAttribute('numGiftCardCaptures', $counts_and_amounts['giftCardCapture']['count']);

        $xml->addAttribute('echeckVerificationAmount', $counts_and_amounts['echeckVerification']['amount']);
        $xml->addAttribute('numEcheckVerification', $counts_and_amounts['echeckVerification']['count']);

        $xml->addAttribute('echeckCreditAmount', $counts_and_amounts['echeckCredit']['amount']);
        $xml->addAttribute('numEcheckCredit', $counts_and_amounts['echeckCredit']['count']);

        $xml->addAttribute('numEcheckRedeposit', $counts_and_amounts['echeckRedeposit']['count']);

        $xml->addAttribute('echeckSalesAmount', $counts_and_amounts['echeckSale']['amount']);
        $xml->addAttribute('numEcheckSales', $counts_and_amounts['echeckSale']['count']);

        $xml->addAttribute('numUpdateCardValidationNumOnTokens', $counts_and_amounts['updateCardValidationNumOnToken']['count']);

        $xml->addAttribute('numUpdateSubscriptions', $counts_and_amounts['updateSubscription']['count']);

        $xml->addAttribute('numCancelSubscriptions', $counts_and_amounts['cancelSubscription']['count']);

        $xml->addAttribute('numCreatePlans', $counts_and_amounts['createPlan']['count']);
        $xml->addAttribute('numUpdatePlans', $counts_and_amounts['updatePlan']['count']);

        $xml->addAttribute('numActivates', $counts_and_amounts['activate']['count']);
        $xml->addAttribute('activateAmount', $counts_and_amounts['activate']['amount']);
        $xml->addAttribute('numDeactivates', $counts_and_amounts['deactivate']['count']);
        $xml->addAttribute('numLoads', $counts_and_amounts['load']['count']);
        $xml->addAttribute('loadAmount', $counts_and_amounts['load']['amount']);
        $xml->addAttribute('numUnloads', $counts_and_amounts['unload']['count']);
        $xml->addAttribute('unloadAmount', $counts_and_amounts['unload']['amount']);
        $xml->addAttribute('numBalanceInquirys', $counts_and_amounts['balanceInquiry']['count']);

        $xml->addAttribute('numAccountUpdates', $counts_and_amounts['accountUpdate']['count']);
        
        $xml->addAttribute('numEcheckPreNoteSale', $counts_and_amounts['echeckPreNoteSale']['count']);
        $xml->addAttribute('numEcheckPreNoteCredit', $counts_and_amounts['echeckPreNoteCredit']['count']);
        
        $xml->addAttribute('submerchantCreditAmount', $counts_and_amounts['submerchantCredit']['amount']);
        $xml->addAttribute('numSubmerchantCredit', $counts_and_amounts['submerchantCredit']['count']);
        $xml->addAttribute('payFacCreditAmount', $counts_and_amounts['payFacCredit']['amount']);
        $xml->addAttribute('numPayFacCredit', $counts_and_amounts['payFacCredit']['count']);
        $xml->addAttribute('reserveCreditAmount', $counts_and_amounts['reserveCredit']['amount']);
        $xml->addAttribute('numReserveCredit', $counts_and_amounts['reserveCredit']['count']);
        $xml->addAttribute('vendorCreditAmount', $counts_and_amounts['vendorCredit']['amount']);
        $xml->addAttribute('numVendorCredit', $counts_and_amounts['vendorCredit']['count']);
        $xml->addAttribute('physicalCheckCreditAmount', $counts_and_amounts['physicalCheckCredit']['amount']);
        $xml->addAttribute('numPhysicalCheckCredit', $counts_and_amounts['physicalCheckCredit']['count']);
        $xml->addAttribute('submerchantDebitAmount', $counts_and_amounts['submerchantDebit']['amount']);
        $xml->addAttribute('numSubmerchantDebit', $counts_and_amounts['submerchantDebit']['count']);
        $xml->addAttribute('payFacDebitAmount', $counts_and_amounts['payFacDebit']['amount']);
        $xml->addAttribute('numPayFacDebit', $counts_and_amounts['payFacDebit']['count']);
        $xml->addAttribute('reserveDebitAmount', $counts_and_amounts['reserveDebit']['amount']);
        $xml->addAttribute('numReserveDebit', $counts_and_amounts['reserveDebit']['count']);
        $xml->addAttribute('vendorDebitAmount', $counts_and_amounts['vendorDebit']['amount']);
        $xml->addAttribute('numVendorDebit', $counts_and_amounts['vendorDebit']['count']);
        $xml->addAttribute('physicalCheckDebitAmount', $counts_and_amounts['physicalCheckDebit']['amount']);
        $xml->addAttribute('numPhysicalCheckDebit', $counts_and_amounts['physicalCheckDebit']['count']);
        $xml->addAttribute('numFundingInstructionVoid', $counts_and_amounts['fundingInstructionVoid']['count']);
        $xml->addAttribute('numFastAccessFunding', $counts_and_amounts['fastAccessFunding']['count']);
        $xml->addAttribute('fastAccessFundingAmount', $counts_and_amounts['fastAccessFunding']['amount']);

        return str_replace("/>", ">", str_replace("<?xml version=\"1.0\"?>\n", "", $xml->asXML()));
    }

    public static function generateRequestHeader($config, $num_batch_requests)
    {
        $xml = simplexml_load_string("<cnpRequest />");

        $xml->addAttribute('numBatchRequests', $num_batch_requests);
        $xml->addAttribute('version', CURRENT_XML_VERSION);
        $xml->addAttribute('xmlns:xmlns','http://www.vantivcnp.com/schema');
        $authentication = $xml->addChild('authentication');
        $authentication->addChild('user',$config["user"]);
        $authentication->addChild('password',$config["password"]);

        return str_replace("<?xml version=\"1.0\"?>\n", "", str_replace("</cnpRequest>", "", $xml->asXML()));
    }

    private static function iterateChildren($data,$transacType)
    {

        foreach ($data as $key => $value) {
            //print $key . " " . $value . "\n";
            if ($value === "REQUIRED") {
                throw new \InvalidArgumentException("Missing Required Field: /$key/");
            } elseif (substr($key, 0, 12) === 'lineItemData') {
                $temp_node = $transacType->addChild('lineItemData');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,-1) == 'detailTax') {
                $temp_node = $transacType->addChild('detailTax');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,14) == 'createDiscount' and $value != null) {
                $temp_node = $transacType->addChild('createDiscount');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,14) == 'updateDiscount') {
                $temp_node = $transacType->addChild('updateDiscount');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,14) == 'deleteDiscount') {
                $temp_node = $transacType->addChild('deleteDiscount');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,11) == 'createAddOn' and $value != null) {
                $temp_node = $transacType->addChild('createAddOn');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,11) == 'updateAddOn') {
                $temp_node = $transacType->addChild('updateAddOn');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,11) == 'deleteAddOn') {
                $temp_node = $transacType->addChild('deleteAddOn');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,-1) == 'lodgingCharge') {
                $temp_node = $transacType->addChild('lodgingCharge');
                Obj2xml::iterateChildren($value,$temp_node);
            } elseif (substr($key,0,16) == 'debitNetworkName') {
                //$temp_node = $transacType->addChild('debitNetworkName');
                $transacType->addChild('debitNetworkName',str_replace('&','&amp;',$value));
            } elseif (((is_string($value)) || is_numeric($value))) {
                $transacType->addChild($key,str_replace('&','&amp;',$value));
            } elseif (is_array($value)) {
                $node = $transacType->addChild($key);
                Obj2xml::iterateChildren($value,$node);
            }
        }
    }

    public static function getConfig($data, $type=NULL)
    {
        $config_array = null;

    $ini_file = realpath(dirname(__FILE__)) . '/cnp_SDK_config.ini';
        if (file_exists($ini_file)) {
            @$config_array =parse_ini_file('cnp_SDK_config.ini');
        }

        if (empty($config_array)) {
            $config_array = array();
        }

        $names = explode(',', CNP_CONFIG_LIST);
        foreach ($names as $name) {
            if (isset($data[$name])) {
                $config[$name] = $data[$name];

            } else {
                if ($name == 'merchantId') {
                    $config['merchantId'] = $config_array['currency_merchant_map']['DEFAULT'];
                } elseif ($name == 'version') {
                    $config['version'] = isset($config_array['version'])? $config_array['version']:CURRENT_XML_VERSION;
                } elseif ($name == 'timeout') {
                        $config['timeout'] = isset($config_array['timeout'])? $config_array['timeout']:'65';
                } else {
                    if ((!isset($config_array[$name])) and ($name != 'proxy')) {
                        throw new \InvalidArgumentException("Missing Field /$name/");
                    }
                    $config[$name] = $config_array[$name];
                }
            }
        }
        if ($type == 'updateSubscription' || $type == 'cancelSubscription' || $type == 'createPlan' || $type == 'updatePlan') {
            if (array_key_exists('reportGroup',$config)) {
                unset($config['reportGroup']);
                $config = array_filter($config);
            }
        }

        return $config;
    }
}
